pkg/services/ngalert/notifier/images.go

```diff
@@ -2,45 +2,162 @@ package notifier
 
 import (
 	"context"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
 	"strings"
 
-	"github.com/grafana/alerting/images"
+	alertingImages "github.com/grafana/alerting/images"
+	alertingModels "github.com/grafana/alerting/models"
+	alertingNotify "github.com/grafana/alerting/notify"
 
+	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 	"github.com/grafana/grafana/pkg/services/ngalert/store"
 )
 
-type imageStore struct {
-	store store.ImageStore
+type imageProvider struct {
+	store  store.ImageStore
+	logger log.Logger
 }
 
-func newImageStore(store store.ImageStore) images.ImageStore {
-	return &imageStore{
-		store: store,
+func newImageProvider(store store.ImageStore, logger log.Logger) alertingImages.Provider {
+	return &imageProvider{
+		store:  store,
+		logger: logger,
 	}
 }
 
-func (i imageStore) GetImage(ctx context.Context, uri string) (*images.Image, error) {
+func (i imageProvider) GetImage(ctx context.Context, uri string) (*alertingImages.Image, error) {
 	var (
 		image *models.Image
 		err   error
 	)
 
 	// Check whether the uri is a URL or a token to know how to query the DB.
 	if strings.HasPrefix(uri, "http") {
+		i.logger.Debug("Received an image URL in annotations")
 		image, err = i.store.GetImageByURL(ctx, uri)
 	} else {
 		token := strings.TrimPrefix(uri, "token://")
+		i.logger.Debug("Received an image token in annotations", "token", token)
 		image, err = i.store.GetImage(ctx, token)
 	}
 	if err != nil {
 		return nil, err
 	}
 
-	return &images.Image{
+	return &alertingImages.Image{
 		Token:     image.Token,
 		Path:      image.Path,
 		URL:       image.URL,
 		CreatedAt: image.CreatedAt,
 	}, nil
 }
+
+func (i imageProvider) GetImageURL(ctx context.Context, alert *alertingNotify.Alert) (string, error) {
+	uri, err := getImageURI(alert)
+	if err != nil {
+		return "", err
+	}
+
+	// If the identifier is a URL, validate that it corresponds to a stored, non-expired image.
+	if strings.HasPrefix(uri, "http") {
+		i.logger.Debug("Received an image URL in annotations", "alert", alert)
+		exists, err := i.store.URLExists(ctx, uri)
+		if err != nil {
+			return "", err
+		}
+		if !exists {
+			i.logger.Warn("Image URL not found in database", "alert", alert)
```

Comment:
I think it can be info or even debug because this event is not something unusual.